triple-lift.js

```diff
@@ -443,29 +601,121 @@ window.headertag.partnerScopes.push(function() {
                     doc.write(ad);
                     doc.close();
                     if (doc.defaultView && doc.defaultView.frameElement) {
                         doc.defaultView.frameElement.width = width;
                         doc.defaultView.frameElement.height = height;
                     }
-                } catch (e){
+                } catch (e) {
                     //? if (DEBUG)
-                    console.log('Error trying to write ' + PARTNER_ID + ' ad to the page');
+                    console.log('Error trying to write ' + PARTNER_ID + ' ad to the page. Error: ' + e);
                 }
 
             }
         };
+
+
+        /* =============================================================================
+             * TL Helpers | Private methods
+             * -----------------------------------------------------------------------------
+             *
+             */
+
+
+        // TODO: Clean this up (break up, redo scope, consider timing)
+        function makeRequest(htSlotId, xSlot, numPlacements, callback) {
+            window.headertag.Network.ajax({
+                url: buildTplCall(xSlot),
+                method: 'GET',
+                partnerId: PARTNER_ID,
+                withCredentials: true,
+                onSuccess: function(response) {
+                    parseResponse(JSON.parse(response));
+                },
+                onFailure: function() {
+                    callback('TPL: demand request failed');
+                }
+            });
+
+            function parseResponse(response) {
+                // TODO validate response
+                var slotDemand = {}
+                if(!response.status) {
+                    var responseSize = response.width + 'x' + response.height;
+                    var creativeStoreId = response.callback_id;
+
+                    creativeStore[creativeStoreId] = creativeStore[creativeStoreId] || {};
+                    creativeStore[creativeStoreId][responseSize] = {};
+                    creativeStore[creativeStoreId][responseSize].ad = response.ad;
+
+                    var demandTargeting = {};
+                    demandTargeting[targetingKeys.idKey] = creativeStoreId;
+                    if (response.deal_id) {
+                        demandTargeting[targetingKeys.pmKey] = responseSize + "_" + bidTransformer.transformBid(response.cpm);
+                        demandTargeting[targetingKeys.pmidKey] = response.deal_id;
+                    }
+                    else {
+                        demandTargeting[targetingKeys.omKey] = responseSize + "_" + bidTransformer.transformBid(response.cpm);
+                    }
+                }
+                slotDemand.timestamp = Utils.now();
+                slotDemand.demand = demandTargeting;
+                demandObj.slot[htSlotId] = slotDemand;
+                // Add timeout to call the callback.
+                if (Object.keys(demandObj.slot).length === numPlacements) {
```

Comment:
As discussed offline, the logic of only calling callback once all bid requests have settled is fine, however, the check itself currently is not fine.

With more thorough testing, we noticed if the number of xSlots configured does not match numPlacements (which is the number of slots available), the callback will never be called.

For example:
If there are 4 placements available and you have a mapping for all 4 placements, then it will be okay.
If there are 5 placements available and you have a mapping for only 4 of the placements, the callback function will never be called.

I would suggest changing numPlacements to the number of slots that were mapped instead of a count to all available slots.